keps/sig-apps/2185-random-pod-select-on-replicaset-downscale/README.md

```diff
@@ -0,0 +1,437 @@
+   Pods that were never restarted, and the remaining domain has 0 Pods.
+   There is nothing to be done here. A random approach would obtain the same
+   result.
+5. The ReplicaSet is upscaled to 3N again, due to user action or HPA
+   recommendation. Due to Pod spreading during scheduling, each domain has N
+   Pods. Balance is recovered. However, one failure domain holds the youngest
+   Pods.
+6. ReplicaSet is downscaled to 2N again. Due to the downscaling preference, all
+   the Pods from one domain are removed, leading to imbalance. The situation
+   doesn't improve with repeated upscale and downscale steps. Instead, a
+   randomized approach leaves about 2/3*N nodes in each
+   failure domain.
+
+### Notes/Constraints/Caveats
+
+The original heuristic, dowscaling the youngest Pods first, has its benefits.
+Newer Pods might not have finished starting up (or warming up) and are likely
+to have less active connections than older Pods. However, this distinction
+doesn't generally apply once Pods have been running steadily for some time.
+
+A purely randomized approach would break those assumptions, potentially
+leading to services disruption. Choosing a heuristic could be left to the user.
+On the other hand, certain workloads take a long time to warm up and, at the
+same time, require high availability.
+
+### Risks and Mitigations
+
+Certain users might be relaying in the existing downscaling heuristic. However,
+there are a number of reasons why we don't need to preserve such behavior as is:
+
+- The behavior is not documented in the API or user docs, only in code.
+- The heuristic is applied last, after other criteria have been applied. In
+  particular, the heuristic doesn't apply when multiple Pods are in the same
+  Node or if a particular Pod has had several container restarts. While users
+  can enforce one Pod per Node with certain scheduling features, there is no
+  workaround for other criteria.
+- We are not proposing to entirely remove the heuristic, just make it more
+  laxed.
+- We are introducing a [related feature](git.k8s.io/enhancements/keps/sig-apps/1828-delete-priority-annotations)
+  that provides higher guarantees for downscaling order that users can migrate
+  to.
+  
+  <!--
+  If not merged yet, the KEP can be found in https://github.com/kubernetes/enhancements/pull/1828
+  -->
+
+## Design Details
+
+We propose a randomized approach to the algorithm for Pod victim selection
+during ReplicaSet downscale:
+
+1. Do a random shuffle of ReplicaSet Pods.
+2. Obtain wall time, and add it to [`ActivePodsWithRanks`](https://github.com/kubernetes/kubernetes/blob/dc39ab2417bfddcec37be4011131c59921fdbe98/pkg/controller/controller_utils.go#L815)
+2. Call sorting algorithm with a modified time comparison for start and
+   creation timestamp.
+
+Instead of directly comparing timestamps, the algorithm compares the elapsed
+times since the timestamp until the current time but in a logarithmic scale,
```

Comment:
> "since the timestamp" 

Which timestamp? creation timestamp?

Reply:
there are 2: creation and ready

Reply:
so we have two sorting keys?

Reply:
I'm confused by this now too, I thought this was just referring to creationTimestamp (picked this part up from Aldo's draft)

Reply:
Diving into implementation details here:
There is not a single sorting key. There are sorting criteria 
https://github.com/kubernetes/kubernetes/blob/cac9339/pkg/controller/controller_utils.go#L784-L809

Two of these criteria are ready time and creation time. We should affect both.
Feel free to include some of these details in the KEP, @damemi

Reply:
right, so criteria 5 and 7 each will be updated so that the key for each is not the timestamp, but the "scale".